### PR TITLE
Cross entropy benchmarking spruce up

### PR DESCRIPTION
**Pull request by mpharrigan**
1. Add an explicit function for generating two-qubit fragments. These are important building blocks for doing "parallel XEB" e.g. for characterization/calibration. The new signature accepts any type of Qid and doesn't require a `pattern` argument.
2. I changed some of the internals of the random circuit generation to use the relevant Cirq type, like `Moment` instead of `Dict[qubit, op]`. 
3. Refactored the docstrings to associate specific elements of the circuit generation to the argument docs that control it. I personally think this is more organized than the wall-o-paragraph approach where every corner case of the function is documented up top
4. Add a notebook that expands upon the mathy docstring of `cirq.experiments.fidelity_estimation.least_squares_xeb_fidelity_from_expectations` with nice diagrams and plots. Let me know if this should be in `docs/` or `docs/xeb/` or `docs/characterization/`.

@XiaoMiQC, who I can't add as reviewer

### CODE REVIEW

File: cirq/experiments/xeb_theory.ipynb [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.plot(fids['cycle_depth'], fids['fid_lsq'], label='LSq')\n",
+    "\n",
+    "xx = np.linspace(0, fids['cycle_depth'].max())\n",
+    "p_depol = 5e-3 # from above\n",
```
Comment (mrwojtek) [inline cirq/experiments/xeb_theory.ipynb:360]: global constant?

**Review by mrwojtek — CHANGES_REQUESTED** [review]

File: cirq/experiments/random_quantum_circuit_generation.py [hunk 2]
```diff
@@ -216,34 +270,23 @@ def random_rotations_between_grid_interaction_layers_circuit(
...
+
         two_qubit_layer = _two_qubit_layer(
             coupled_qubit_pairs, two_qubit_op_factory, pattern[i % len(pattern)], prng
         )
-        circuit.append(
-            [g.on(q) for q, g in single_qubit_layer.items()],
-            strategy=circuits.InsertStrategy.NEW_THEN_INLINE,
-        )
-        circuit.append(two_qubit_layer, strategy=circuits.InsertStrategy.EARLIEST)
+        circuit.append(two_qubit_layer)
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/random_quantum_circuit_generation.py:285]: yup! changing to be consistent with `+= single_qubit_layer`.

File: cirq/experiments/random_quantum_circuit_generation_test.py [hunk 3]
```diff
@@ -22,10 +22,53 @@
...
+
+    circuit = random_rotations_between_two_qubit_circuit(
+        q0, q1, 4, seed=52, add_final_single_qubit_layer=False
+    )
+    assert len(circuit) == 4 * 2
+    assert circuit.all_qubits() == {q0, q1}
+
+    print()
+    print('-' * 20)
+    print(circuit.to_text_diagram(transpose=True))
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/random_quantum_circuit_generation_test.py:46]: yes, sorry

File: cirq/experiments/xeb_theory.ipynb [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Make long circuits (which we will truncate)\n",
+    "circuits = [\n",
+    "    rqcg.random_rotations_between_two_qubit_circuit(\n",
+    "        q0, q1, \n",
+    "        depth=100, \n",
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb:144]: done

File: cirq/experiments/xeb_theory.ipynb [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pure_sim = cirq.Simulator()\n",
+    "noisy_sim = cirq.DensityMatrixSimulator(noise=cirq.depolarize(5e-3))\n",
+    "\n",
+    "records = []\n",
+    "for cycle_depth in cycle_depths:\n",
+    "    for circuit_i, circuit in enumerate(circuits):\n",
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb:181]: I'm on the fence about this one. Using itertools does reduce the maximum line length for this cell, but only slightly since the itertools line is so long itself. I think the nested for loop may be more pedagogical. What do you think?

I'll re-write the `noisy_sim.sample(...` line anyways to be less long

File: cirq/experiments/xeb_theory.ipynb [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    "        tcircuit = circuit[:circuit_depth]\n",
+    "\n",
+    "        # Pure-state simulation\n",
+    "        psi = pure_sim.simulate(tcircuit)\n",
+    "        psi = psi.final_state_vector\n",
+    "        pure_probs = np.abs(psi)**2\n",
+    "\n",
+    "        # Noisy execution\n",
+    "        sampled_inds = noisy_sim.sample(tcircuit + cirq.measure(q0, q1), repetitions=10_000).values[:,0]\n",
+    "        sampled_probs = np.bincount(sampled_inds, minlength=2**2) / len(sampled_inds)\n",
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb:195]: extracted as a variable with a comment briefly explaining

File: cirq/experiments/xeb_theory.ipynb [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    "import seaborn as sns\n",
+    "colors = sns.cubehelix_palette(n_colors=len(cycle_depths))        \n",
+    "colors = {k: colors[i] for i, k in enumerate(cycle_depths)}\n",
+    "\n",
+    "def per_cycle_depth(df):\n",
+    "    fid_lsq = df['numerator'].sum() / df['denominator'].sum()\n",
+    "    \n",
+    "    cycle_depth = df.name    \n",
+    "    xx = np.linspace(0, df['x'].max())\n",
+    "    plt.plot(xx, fid_lsq*xx, color=colors[cycle_depth])\n",
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb:336]: done. Added a legend too

File: cirq/experiments/xeb_theory.ipynb (hunk 1 above)
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb:360]: done

**Comment by mpharrigan** [comment]
@balopat I see there's a check that means I can't use `cirq.contrib` in notebooks; specifically `cirq.contrib.svg.SVGCircuit`. I think this is a really important utility to make rich notebooks with pretty circuit diagrams. Should we revisit this policy? Move SVG out of contrib? I don't necessarily want this to block the current PR

**Comment by mpharrigan** [comment]
I'm also somewhat stumped by the notebook check error, per usual

File: cirq/experiments/xeb_theory.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (mrwojtek) [reply cirq/experiments/xeb_theory.ipynb:187]: Yeah, that seems reasonable; keeping nested loops like that is easier to read.

File: cirq/experiments/xeb_theory.ipynb [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.plot(fids['cycle_depth'], fids['fid_lsq'], label='LSq')\n",
+    "\n",
+    "xx = np.linspace(0, fids['cycle_depth'].max())\n",
+    "plt.plot(xx, (1-P_DEPOL)**(4*xx), label=r'$(1-\\mathrm{depol})^{4d}$')\n",
+    "\n",
+    "plt.ylabel('Circuit fidelity', fontsize=18)\n",
+    "plt.xlabel('Cycle Depth', fontsize=18)\n",
```
Comment (mrwojtek) [inline cirq/experiments/xeb_theory.ipynb:377]: 'Cycle dept d' to connect with {4d} in the label above?

**Review by mrwojtek — APPROVED** [review]

File: cirq/experiments/xeb_theory.ipynb (hunk 8 above)
Comment (mpharrigan) [reply cirq/experiments/xeb_theory.ipynb]: done

**Comment by balopat** [comment]
> I'm also somewhat stumped by the notebook check error, per usual

I'm sorry you have to deal with this :/ #3603

**Comment by mpharrigan** [comment]
I can't reproduce the error when running ` pytest -m slow dev_tools/notebook_test.py  -v` locally

**Comment by mpharrigan** [comment]
However, when running locally I run into the issue that this notebook tries to use a function that was introduced in this PR, but it installs the stable version of `cirq` from pip, which obviously doesn't work. Do I have to split this into two PRs and then use `cirq-nightly` in the notebook? I feel like this issue will come up a lot if we want people to document new features with notebooks as they contribute them

**Comment by balopat** [comment]
> However, when running locally I run into the issue that this notebook tries to use a function that was introduced in this PR, but it installs the stable version of `cirq` from pip, which obviously doesn't work. Do I have to split this into two PRs and then use `cirq-nightly` in the notebook? I feel like this issue will come up a lot if we want people to document new features with notebooks as they contribute them

This is a general issue with notebooks: they are targeted at the latest stable version. We don't have a concept of "nightly notebooks" yet, but this could be one of the first ones. We also don't test notebook compatibility with master either. We can start to think about this. 

For now, as this is a new notebook, and it is not part of the devsite documentation yet, I would just exclude it explicitly from the notebook_tests (the ignore pattern at the top of the file), make sure that the notebook actually does install `cirq --pre` and put in a warning that this notebook is based on unreleased features.

**Comment by mpharrigan** [comment]
done.

Really, the notebooks should be versioned like Cirq proper. Ideally we should have the notebook tests run against cirq-as-in-the-pr (auto-installed in environment) and publish a colab environment that has cirq pre-installed. I took a look at the tensorflow examples and they don't install tensorflow every time for each notebook

**Comment by mpharrigan** [comment]
PTAL code owners

File: cirq/experiments/random_quantum_circuit_generation.py [hunk 9]
```diff
@@ -154,6 +154,69 @@ def __repr__(self) -> str:
 )
 
 
+def random_rotations_between_two_qubit_circuit(
```
Comment (balopat) [inline cirq/experiments/random_quantum_circuit_generation.py:157]: wow, this method name is tough. I came up with even longer ones :)
- `random_rotation_layers_interleaved_with_two_qubit_gate_layers_circuit`
- `random_layered_interleaved_single_and_two_qubit_gates_circuit`
- `alternating_random_single_and_two_qubit_layers_circuit`

Comment (balopat) [reply cirq/experiments/random_quantum_circuit_generation.py:157]: (optional)

**Review by balopat — APPROVED** [review]
Approved with nits - I'm going to review the notebook too

File: docs/characterization/xeb_theory.ipynb [hunk 10]
```diff
@@ -0,0 +1,454 @@
+{
```
Comment (balopat) [inline docs/characterization/xeb_theory.ipynb:1]: <p>nit: expectation <em>value</em> of $O_U$</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/3669/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (balopat) [inline docs/characterization/xeb_theory.ipynb:1]: <p>are we assuming people have seaborn installed? I guess on colab this is true. We should figure out how we handle requirements for notebooks. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/quantumlib/Cirq/pull/3669/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

File: cirq/experiments/random_quantum_circuit_generation.py (hunk 9 above)
Comment (mpharrigan) [reply cirq/experiments/random_quantum_circuit_generation.py:157]: the current name was by analogy to the existing function `random_rotations_between_grid_interaction_layers_circuit`. Let's circle back on this name later. End users likely won't be using this directly, and I think the analogy to the other function is helpful for developers. This notebook shows the details that will be abstracted away in follow on PRs.

File: docs/characterization/xeb_theory.ipynb (hunk 10 above)
Comment (mpharrigan) [reply docs/characterization/xeb_theory.ipynb:1]: <p>done</p>

Comment (mpharrigan) [reply docs/characterization/xeb_theory.ipynb:1]: <p>it's in <code>notebook_test.py</code>  as an installed package / requirement. I assumed other notebooks were therefore using it</p>

**Comment by mpharrigan** [comment]
@balopat back at you

File: docs/characterization/xeb_theory.ipynb (hunk 10 above)
Comment (balopat) [reply docs/characterization/xeb_theory.ipynb:1]: Oh, you're right. I added those there as an assumption that colab has seaborn installed.

File: cirq/experiments/random_quantum_circuit_generation.py (hunk 9 above)
Comment (balopat) [reply cirq/experiments/random_quantum_circuit_generation.py:157]: sgtm!
